onnxruntime/contrib_ops/cpu/bert/multihead_attention_helper.h

```diff
@@ -178,7 +225,8 @@ Status CheckInputs(const T* query,
         return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "Input 'value' shall be 4D when 'key' is 4D");
       }
 
-      if (bias != nullptr) {
+      // Bias is supported in DecoderMaskedMultiHeadAttention only for cross attention.
```

Comment:
only for self-attn? k_bias and v_bias are added to the k and v under the condition !is_cross_attn(https://github.com/microsoft/onnxruntime/blob/4167b68abf2715c52439874ad9ddeaebfb3dafcb/onnxruntime/contrib_ops/cuda/bert/fastertransformer_decoder_attention/decoder_masked_multihead_attention_impl.cu#L197)

Reply:
This code path is for cross attention. Bias is not supported in MultiHeadAttention for cross attention, but possible for DecoderMaskedMultiHeadAttention. There is some assumption in DecoderMaskedMultiHeadAttention implementation, that there is no bias for k/v; and only q might have bias for cross attention.

Reply:
looks like bias is supported in both self and cross attn in this PR: https://github.com/microsoft/onnxruntime/pull/15827

Reply:
> Bias is not supported in MultiHeadAttention for cross attention, but possible for DecoderMaskedMultiHeadAttention.

It is supported for `MultiHeadAttention` on CPU but the `assert(bias == nullptr)` should be removed.

https://github.com/microsoft/onnxruntime/blob/278f0f5cd2ab18abb48f381fc3fee127ebe960d2/onnxruntime/contrib_ops/cpu/bert/multihead_attention.cc#L132-L153

It is supported for `MultiHeadAttention` on CUDA in the unfused attention case.

https://github.com/microsoft/onnxruntime/blob/278f0f5cd2ab18abb48f381fc3fee127ebe960d2/onnxruntime/contrib_ops/cuda/bert/attention_prepare_qkv.cu#L195-L223

It is also supported for `DecoderMaskedMultiHeadAttention` on CUDA for when `MultiHeadAttention` is replaced with `DecoderMaskedMultiHeadAttention` under the following conditions.
- CUDA EP
- FP16 precision
- Past-present share buffer = true

This case is covered in the above three kernels because it happens in the decoder component of the OpenAI Whisper model for the following scenario.
- Query is of shape `(B, S, D)` and has to perform `AddBiasTranspose` before computing attention.
- Both key and value are of shape `(B, N, S, H)` and do not need to perform `AddBiasTranspose` because they are past key-value caches used as the current key-value inputs and passed directly to attention with the right shape.
- The bias is right-padded with 0s in the below picture for the key and value in case the kernel selected does not support adding just the query portion of the bias.

![image](https://github.com/user-attachments/assets/66b22319-a10c-4ae4-b092-013af2cc3d03)